File: ios/Sources/VideoRecorder/TranscodeSettings.swift
import Foundation

enum TranscodeSettingsError: Error {
    case invalidArgument(message: String)
}

/// Output video settings of an edit operation.
public class TranscodeSettings: NSObject {
    private var height: Int = 0
    private var width: Int = 0
    private var keepAspectRatio: Bool = true
    private var fps: Int = 30
    private var videoBitrate: Int = 0

    init(height: Int, width: Int, keepAspectRatio: Bool, fps: Int, videoBitrate: Int = 0) throws {
        super.init()

        try self.setHeight(height)
        try self.setWidth(width)
        self.setKeepAspectRatio(keepAspectRatio)
        try self.setFps(fps)
        try self.setVideoBitrate(videoBitrate)
    }

    func getHeight() -> Int {
        return self.height
    }

    func setHeight(_ height: Int) throws {
        if height < 0 {
            throw TranscodeSettingsError.invalidArgument(message: "Parameter height cannot be negative")
        }
        self.height = height
    }

    func getWidth() -> Int {
        return self.width
    }

    func setWidth(_ width: Int) throws {
        if width < 0 {
            throw TranscodeSettingsError.invalidArgument(message: "Parameter width cannot be negative")
        }
        self.width = width
    }

    func isKeepAspectRatio() -> Bool {
        return self.keepAspectRatio
    }

    func setKeepAspectRatio(_ keepAspectRatio: Bool) {
        self.keepAspectRatio = keepAspectRatio
    }

    func getFps() -> Int {
        return self.fps
    }

    func setFps(_ fps: Int) throws {
        if fps < 1 {
            throw TranscodeSettingsError.invalidArgument(message: "Parameter fps cannot be lower than 1")
        }
        self.fps = fps
    }

    /// Requested output bitrate in bits/sec, or 0 to let the plugin estimate one.
    func getVideoBitrate() -> Int {
        return self.videoBitrate
    }

    func setVideoBitrate(_ videoBitrate: Int) throws {
        if videoBitrate < 0 {
            throw TranscodeSettingsError.invalidArgument(message: "Parameter videoBitrate cannot be negative")
        }
        self.videoBitrate = videoBitrate
    }
}

File: android/src/main/java/com/capacitorcommunity/videorecorder/editor/TranscodeSettings.java
package com.capacitorcommunity.videorecorder.editor;

/**
 * Output video settings of an edit operation.
 */
public class TranscodeSettings {

    private int height = 0;
    private int width = 0;
    private boolean keepAspectRatio = true;
    private int fps = 30;
    private int videoBitrate = 0;

    public TranscodeSettings() {}

    public TranscodeSettings(int height, int width, boolean keepAspectRatio, int fps) {
        this(height, width, keepAspectRatio, fps, 0);
    }

    public TranscodeSettings(int height, int width, boolean keepAspectRatio, int fps, int videoBitrate) {
        setHeight(height);
        setWidth(width);
        setKeepAspectRatio(keepAspectRatio);
        setFps(fps);
        setVideoBitrate(videoBitrate);
    }

    public int getHeight() {
        return height;
    }

    public void setHeight(int height) {
        if (height < 0) {
            throw new IllegalArgumentException("Parameter height cannot be negative");
        }

        this.height = height;
    }

    public int getWidth() {
        return width;
    }

    public void setWidth(int width) {
        if (width < 0) {
            throw new IllegalArgumentException("Parameter width cannot be negative");
        }

        this.width = width;
    }

    public boolean isKeepAspectRatio() {
        return keepAspectRatio;
    }

    public void setKeepAspectRatio(boolean keepAspectRatio) {
        this.keepAspectRatio = keepAspectRatio;
    }

    public int getFps() {
        return fps;
    }

    public void setFps(int fps) {
        if (fps < 1) {
            throw new IllegalArgumentException("Parameter fps cannot be lower than 1");
        }

        this.fps = fps;
    }

    /**
     * @return the requested output bitrate in bits/sec, or 0 to let the plugin estimate one
     */
    public int getVideoBitrate() {
        return videoBitrate;
    }

    public void setVideoBitrate(int videoBitrate) {
        if (videoBitrate < 0) {
            throw new IllegalArgumentException("Parameter videoBitrate cannot be negative");
        }

        this.videoBitrate = videoBitrate;
    }
}
